File: Sources/NiuLaiMarketPets/NotificationPolicy.swift
import Foundation

/// Decides whether a settled automatic pet transition is notification-worthy.
/// The state engine owns debounce/cooldown; this policy owns notification
/// semantics and suppresses repeated boundary events.
public struct NotificationPolicy: Sendable {
    private var lastNotifiedTransition: String?

    public init() {}

    public static func title(for pet: PetID) -> String {
        switch pet {
        case .muamua: return "妈妈——"
        case .baola: return "豹拉！"
        case .niulai: return "牛来了"
        }
    }

    public static func body(for pet: PetID, target: MarketTarget, quote: Quote?) -> String {
        "\(pet.displayName) · \(target.name) · \(MarketRules.price(quote?.lastPrice)) · \(MarketRules.signedPercent(quote?.percent))"
    }

    public mutating func reset() {
        lastNotifiedTransition = nil
    }

    public mutating func shouldNotify(
        previousPercent: Double?,
        currentPercent: Double,
        switchedTo: PetID,
        mode: ControlMode,
        isInitialSample: Bool,
        isTargetRotation: Bool,
        isStale: Bool
    ) -> Bool {
        guard mode == .auto,
              !isInitialSample,
              !isTargetRotation,
              !isStale,
              currentPercent.isFinite,
              let previousPercent,
              previousPercent.isFinite,
              let currentBucket = MarketRules.bucket(for: currentPercent),
              currentBucket == switchedTo,
              let previousBucket = MarketRules.bucket(for: previousPercent),
              previousBucket != currentBucket
        else { return false }

        let transition = "\(previousBucket.rawValue)->\(currentBucket.rawValue)"
        guard transition != lastNotifiedTransition else { return false }
        lastNotifiedTransition = transition
        return true
    }
}

File: platforms/windows/Core/NotificationPolicy.cs
namespace NiuLaiMarketPets.Windows.Core;

public sealed class NotificationPolicy
{
    private string? _lastTransition;

    public void Reset() => _lastTransition = null;

    public bool ShouldNotify(double? previousPercent, double currentPercent, PetId switchedTo, ControlMode mode, bool initialSample, bool targetRotation, bool stale)
    {
        if (mode != ControlMode.Auto || initialSample || targetRotation || stale || !double.IsFinite(currentPercent) || previousPercent is null || !double.IsFinite(previousPercent.Value)) return false;
        var previous = MarketRules.Bucket(previousPercent.Value);
        var current = MarketRules.Bucket(currentPercent);
        if (previous is null || current is null || previous == current || current != switchedTo) return false;
        var transition = $"{previous}->{current}";
        if (transition == _lastTransition) return false;
        _lastTransition = transition;
        return true;
    }

    public static string Title(PetId pet) => pet switch
    {
        PetId.Muamua => "妈妈——",
        PetId.Baola => "豹拉！",
        _ => "牛来了"
    };

    public static string Body(PetId pet, MarketTarget target, Quote? quote) =>
        $"{pet.DisplayName()} · {target.Name} · {MarketRules.Price(quote?.LastPrice)} · {MarketRules.SignedPercent(quote?.Percent)}";
}
